lxd/api_metrics.go

```diff
@@ -323,31 +331,31 @@ func metricsGet(d *Daemon, r *http.Request) response.Response {
 }
 
 func getFilteredMetrics(s *state.State, r *http.Request, compress bool, metricSet *metrics.MetricSet) response.Response {
+	// Ignore filtering in case the authentication for metrics is disabled.
 	if !s.GlobalConfig.MetricsAuthentication() {
 		return response.SyncResponsePlain(true, compress, metricSet.String())
 	}
 
-	// Get instances the user is allowed to view.
-	userHasPermission, err := s.Authorizer.GetPermissionChecker(r.Context(), r, auth.EntitlementCanView, entity.TypeInstance)
-	if err != nil && !auth.IsDeniedError(err) {
+	userHasProjectPermission, err := s.Authorizer.GetPermissionChecker(r.Context(), r, auth.EntitlementCanViewMetrics, entity.TypeProject)
+	if err != nil {
 		return response.SmartError(err)
-	} else if err != nil {
-		// This is counterintuitive. We are unauthorized to get a permission checker for viewing instances because a metric type certificate
-		// can't view instances. However, in order to get to this point we must already have auth.EntitlementCanViewMetrics. So we can view
-		// the metrics but we can't do any filtering, so just return the metrics.
-		return response.SyncResponsePlain(true, compress, metricSet.String())
 	}
 
-	// Filter by project name and instance name.
-	metricSet.FilterSamples(userHasPermission)
-
-	userHasPermission, err = s.Authorizer.GetPermissionChecker(r.Context(), r, auth.EntitlementCanView, entity.TypeProject)
+	userHasServerPermission, err := s.Authorizer.GetPermissionChecker(r.Context(), r, auth.EntitlementCanViewMetrics, entity.TypeServer)
```

Comment:
Hmm so getting a permission checker for the server entity type is something I hadn't anticipated. 

When getting a permission checker, the OpenFGA driver asks "which objects of type server is the caller related to via the given relation?". The driver returns a function which compares the input URL to the list of objects returned by OpenFGA.

However, for us there is only one entity of type server, so the list will either have one entry (`server:/1.0`) or none. Therefore I had expected that for the server entity type, we would always be calling `CheckPermission` instead. I'm glad that this is working but this usage isn't something I had considered in the OpenFGA datastore (database layer).

It's not clear from the `Authorizer` interface that this isn't expected so I'm not suggesting a change here, but perhaps it would be a good idea to add something to the embedded OpenFGA `Authorizer`. For example:

```golang
// There is only one server entity, so no need to do a ListObjects request. Instead perform a permission check against
	// the server URL and return an appropriate PermissionChecker.
	if entityType == entity.TypeServer {
		err = e.CheckPermission(r.Context(), r, entity.ServerURL(), entitlement)
		if err == nil {
			return allowFunc(true), nil
		} else if IsDeniedError(err) {
			return allowFunc(false), nil
		} else {
			return nil, fmt.Errorf("Failed to get a permission checker: %w", err)
		}
	}
```

Reply:
Yes good idea

Reply:
Covered by https://github.com/canonical/lxd/issues/13236.